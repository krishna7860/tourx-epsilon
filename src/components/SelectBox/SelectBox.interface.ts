export interface OptionProps {
  _id: number;
  title: string;
  slug: string;
  createdAt: string;
  name?: string;
  stateId?: string;
}

export interface SelectBoxProps {
  placeholder?: string;
  value?: string;
  options: OptionProps[];
  type?: string;
  onChange: (params: string) => void;
  disabled?: boolean;
}
