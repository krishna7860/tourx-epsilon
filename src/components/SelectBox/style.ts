import styled from "styled-components";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";

export const Container = styled.div`
  margin: 0.5rem 0.75rem;
  position: relative;
  cursor: pointer;
`;

export const Input = styled.input`
  padding: 0.75rem 0.75rem;
  padding-right: 2.5rem;
  width: 100%;
  border: 1px solid #cfd0d7;
  color: black;
  outline: none;
  border-radius: 8px;
  cursor: pointer;
  ::placeholder {
    color: #cfd0d7;
    font-size: 14px;
  }
  :focus {
    transition: all 0.1s;
    border: 2px solid #d5d8dc;
  }
`;

export const Droplist = styled.div`
  position: absolute;
  z-index: 3;
  width: 100%;
  background: white;
  max-height: 300px;
  overflow-y: auto;
  box-shadow: 0 3px 6px -4px rgb(0 0 0 / 12%), 0 6px 16px 0 rgb(0 0 0 / 8%),
    0 9px 28px 8px rgb(0 0 0 / 5%);
`;

export const Options = styled.div`
  padding: 0.75rem 0.75rem;
  width: 100%;
  font-size: 14px;
  cursor: pointer;
  :hover {
    background: #f2f4f4;
  }
  z-index: 5;
`;

export const DropArrow = styled(ExpandMoreIcon)`
  position: absolute;
  top: 20%;
  right: 3%;
`;
