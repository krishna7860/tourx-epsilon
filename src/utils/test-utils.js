/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/prop-types */
import React from "react";
import { render as rtlRender } from "@testing-library/react";
// import { renderHook as rtlRenderHooks } from "@testing-library/react-hooks";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import * as themes from "../theme/schema.json";
import reducer from "../redux/rootReducer";
// import defaultStore from "../redux/defaultStore";

function render(ui, options = {}) {
  const {
    initialState = {},
    store = createStore(reducer, initialState),
    theme,
    ...renderOptions
  } = options;

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <ThemeProvider theme={themes.data.light}>{children}</ThemeProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// function renderHookWithRedux(callback, options = {}) {
//   const { initialState = defaultStore, store = createStore(reducer, initialState), ...others } = options

//   const returned = rtlRenderHooks(callback, {
//     ...others,
//     // eslint-disable-next-line react/display-name
//     wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
//   })

//   return {
//     ...returned,
//     store
//   }
// }

// re-export everything
export * from "@testing-library/react";
// override render method
export { render, rtlRender as renderHook };
