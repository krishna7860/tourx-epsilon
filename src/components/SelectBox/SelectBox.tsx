/* eslint-disable react-hooks/exhaustive-deps */
// eslint-disable-next-line react/no-unused-prop-types
import React, { useEffect } from "react";
import useGenericState from "../../Library/useGenericState";
import { OptionProps, SelectBoxProps } from "./SelectBox.interface";
import { Input, Droplist, Container, Options, DropArrow } from "./style";

const SelectBox = (props: SelectBoxProps): JSX.Element => {
  const { placeholder, value, options, onChange, disabled } = props;
  const [state, setState] = useGenericState({
    showDropdownList: false,
    currentValue: value || "",
    dropdownList: options,
  });

  useEffect(() => {
    setState({ dropdownList: options });
  }, [options, value]);

  const { showDropdownList, currentValue, dropdownList } = state;

  const handleInputChange = (selected: string) => {
    setState(
      { currentValue: selected, showDropdownList: false },
      (newState) => {
        onChange(newState.currentValue);
      }
    );
  };

  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>): boolean => {
    if (e.target.value === "") {
      setState({ currentValue: e.target.value, dropdownList: options });
      return true;
    }
    let datalist = options;
    datalist = datalist.filter((item) =>
      item.slug
        .toLocaleLowerCase()
        .startsWith(e.target.value.toLocaleLowerCase())
    );
    setState({ currentValue: e.target.value, dropdownList: datalist });
    return true;
  };

  return (
    <Container className="selectBox">
      <Input
        placeholder={placeholder}
        value={currentValue}
        onChange={handleSearch}
        onClick={() => setState({ showDropdownList: !state.showDropdownList })}
        disabled={disabled}
        autoFocus
      />
      {showDropdownList && (
        <Droplist>
          {dropdownList.map((option: OptionProps) => (
            <Options
              key={option._id}
              data-value={option.slug}
              data-test="select-options"
              onClick={() => {
                handleInputChange(option.title ?? option.name);
              }}
            >
              {option.title ?? option.name}
            </Options>
          ))}
        </Droplist>
      )}
      <DropArrow
        onClick={() => setState({ showDropdownList: !state.showDropdownList })}
      />
    </Container>
  );
};

export default SelectBox;
