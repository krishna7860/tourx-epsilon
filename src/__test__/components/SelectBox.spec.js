/* eslint-disable react/jsx-filename-extension */
import React from "react";
import { render } from "../../utils/test-utils";
import SelectBox from "../../components/SelectBox/SelectBox";

describe("components/SelectBox", () => {
  const props = {
    type: "input",
    placeholder: "Please Select State",
  };
  it("should be rendered properly", () => {
    const { asFragment } = render(
      <SelectBox
        type={props.type}
        placeholder={props.placeholder}
        options={[]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
