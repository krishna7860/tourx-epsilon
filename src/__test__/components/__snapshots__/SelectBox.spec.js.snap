// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SelectBox should be rendered properly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM hLnSep selectBox"
  >
    <input
      class="sc-gtsrHT eXYDDY"
      placeholder="Please Select State"
      value=""
    />
  </div>
</DocumentFragment>
`;
